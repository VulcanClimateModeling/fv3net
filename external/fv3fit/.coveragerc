[run]
omit = tests/* setup.py fv3fit/sklearn/__main__.py fv3fit/keras/__main__.py
