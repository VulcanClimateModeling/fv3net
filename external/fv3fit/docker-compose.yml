# this is only for development purposes and integration with visual studio code
version: '3.4'
services:
  fv3fit:
    build:
      context: "../../"
      # dockerfile is relative to context
      dockerfile: "docker/fv3fit/Dockerfile"
      target: bld
    volumes:
      - ../../external/vcm:/external/vcm
      - ../../external/loaders:/external/loaders
      - ../../external/synth:/external/synth
      - .:/fv3fit
