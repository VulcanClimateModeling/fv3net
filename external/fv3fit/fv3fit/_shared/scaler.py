import abc
import numpy as np
from typing import Sequence, Mapping, Union, BinaryIO


class NormalizeTransform(abc.ABC):
    @abc.abstractmethod
    def fit():
        pass

    @abc.abstractmethod
    def normalize(y: np.ndarray):
        pass
    
    @abc.abstractmethod
    def denormalize(y: np.ndarray):
        pass

    @abc.abstractmethod
    def dump():
        pass

    @abc.abstractmethod
    def load():
        pass


class StandardScaler(NormalizeTransform):
    def __init__(self):
        self.mean = None
        self.std = None

    def fit(self, data: np.ndarray):
        self.mean = data.mean(axis=0).astype(np.float32)
        self.std = data.std(axis=0).astype(np.float32)

    def normalize(self, data):
        return (data - self.mean) / self.std

    def denormalize(self, data):
        return data * self.std + self.mean

    def dump(self, f: BinaryIO):
        data = {}
        if self.mean is not None:
            data["mean"] = self.mean
        if self.std is not None:
            data["std"] = self.std
        return np.savez(f, **data)

    @classmethod
    def load(cls, f: BinaryIO):
        data = np.load(f)
        scaler = cls()
        scaler.mean = data.get("mean")
        scaler.std = data.get("std")
        return scaler


class MassScaler(NormalizeTransform):
    def __init__(self):
        self.weights = None

    def fit(
            self,
            output_var_order: Sequence[str],
            output_var_feature_count: Mapping[str, int],
            delp_weights: np.ndarray,
            variable_scale_factors: Mapping[str, float] = None,
            sqrt_weights: bool = False
    ):
        """Weights vertical variables by their relative masses (via delp)
        and upscales variables by optional scale factors.

        Args:
            output_var_order: Sequence of target variable names
            output_var_feature_count: Dict with number of features in each variable.
                2D variables have count 1, 3D should have number of features equal
                to number of vertical levels.
            delp_weights: 1D array of pressure thicknesses for each model level, used to
                scale weights by the levels' relative masses.
            variable_scale_factors: Optional mapping of variable names to scale factors
                by which their weights will be multiplied when normalizing. This allows the mass
                weighted outputs to be scaled to the same order of magnitude.
                Default of None will target dQ2 features by a factor of 1000.
            sqrt_weights: If True, will square root the weights returned by
                _create_weight_array. Useful if this is used as a target transform
                regressor in fv3fit.sklearn with a MSE loss, as there is no current way
                to directly weight the loss function terms. If set to take sqrt of
                weights in the target transform, the MSE loss function terms will be
                approximately weighted to the layer mass.
        """
        self._variable_scale_factors = variable_scale_factors or {"dQ2": 1000.}
        self.weights = self._create_weight_array(
            delp_weights, output_var_order, output_var_feature_count)
        if sqrt_weights:
            self.weights = np.sqrt(self.weights)

    def _create_weight_array(
            self,
            delp_weights: np.ndarray,
            output_var_order: Sequence[str],
            output_var_feature_count: Mapping[str, int]):
        n_levels = len(delp_weights)
        weights = np.array([])
        for var in output_var_order:
            n_features = output_var_feature_count[var]
            if n_features == n_levels:
                var_weights = delp_weights
            elif n_features == 1:
                var_weights = np.array([1.])
            else:
                raise ValueError(
                    f"Output variable {var} has {n_features} features > 1 "
                    f"but not equal to number of vertical levels {n_levels}.")
            if var in self._variable_scale_factors:
                # want to multiply by scale factor when dividing by weights
                var_weights /= self._variable_scale_factors[var]
            np.append(weights, var_weights)
        return weights
    
    def normalize(self, y: np.ndarray):
        return y / self.weights

    def denormalize(self, y: np.ndarray):
        return y * self.weights

    def dump(self, f: BinaryIO):
        data = {}
        if self.weights is not None:
            data["weights"] = self.weights
        return np.savez(f, **data)

    @classmethod
    def load(cls, f: BinaryIO):
        data = np.load(f)
        scaler = cls()
        scaler.weights = data.get("weights")
        return scaler
    

        