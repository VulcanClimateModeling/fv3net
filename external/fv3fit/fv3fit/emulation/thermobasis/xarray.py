import dataclasses
import logging
from typing import MutableMapping, Hashable, Union
import xarray as xr
import tensorflow as tf
from fv3fit.emulation.thermobasis.batch import to_tensors, to_dict_no_static_vars
from fv3fit.emulation.thermobasis.emulator import Trainer, Config

State = MutableMapping[Hashable, xr.DataArray]


@dataclasses.dataclass
class XarrayEmulator:
    """Wrap an OnlineEmulator to allow for xarrray inputs and outputs
    """

    emulator: Trainer

    @property
    def input_variables(self):
        return self.emulator.input_variables

    @property
    def output_variables(self):
        return self.emulator.output_variables

    def dump(self, path: str):
        return self.emulator.dump(path)

    @classmethod
    def load(cls, path: str):
        return cls(Trainer.load(path))

    def predict(self, state: State) -> State:
        in_ = stack(state, self.input_variables)
        in_tensors = to_tensors(in_)
        x = self.emulator.batch_to_specific_humidity_basis(in_tensors)
        out = self.emulator.model(x)

        tensors = to_dict_no_static_vars(out)

        dims = ["sample", "z"]
        attrs = {"units": "no one cares"}

        return dict(
            xr.Dataset(
                {key: (dims, val, attrs) for key, val in tensors.items()},
                coords=in_.coords,
            ).unstack("sample")
        )

    def partial_fit(self, statein: State, stateout: State):

        in_tensors = _xarray_to_tensor(statein, self.input_variables)
        out_tensors = _xarray_to_tensor(stateout, self.output_variables)
        d = tf.data.Dataset.from_tensor_slices((in_tensors, out_tensors)).shuffle(
            1_000_000
        )
        self.emulator.batch_fit(d)


def _xarray_to_tensor(state, keys):
    in_ = stack(state, keys)
    return to_tensors(in_)


def stack(state: State, keys) -> xr.Dataset:
    ds = xr.Dataset({key: state[key] for key in keys})
    sample_dims = ["y", "x"]
    return ds.stack(sample=sample_dims).transpose("sample", ...)


def get_xarray_emulator(config: Union[Config, str]) -> XarrayEmulator:
    if isinstance(config, str):
        logging.info(f"Loading emulator from checkpoint {config}")
        return XarrayEmulator(Trainer.load(config))
    else:
        return XarrayEmulator(Trainer(config))
