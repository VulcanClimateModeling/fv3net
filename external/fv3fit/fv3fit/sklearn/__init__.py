from ._wrapper import SklearnWrapper, RegressorEnsemble
