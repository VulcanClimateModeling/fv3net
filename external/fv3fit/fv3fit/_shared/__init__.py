from .config import (
    TrainingConfig,
    register_training_function,
)
from .packer import (
    pack,
    unpack,
    count_features,
    ArrayPacker,
    unpack_matrix,
    multiindex_to_tuple,
    tuple_to_multiindex,
)
from .scaler import (
    StandardScaler,
    ManualScaler,
    get_mass_scaler,
    get_scaler,
    NormalizeTransform,
)
from .predictor import Predictor
from .stacking import (
    StackedBatches,
    stack_non_vertical,
    stack,
    match_prediction_to_input_coords,
    SAMPLE_DIM_NAME,
)
from .models import EnsembleModel, DerivedModel
from .filesystem import get_dir, put_dir
