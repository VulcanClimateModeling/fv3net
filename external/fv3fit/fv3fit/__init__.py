from ._shared.predictor import Predictor, Estimator
from ._shared.io import dump, load
from . import keras
from . import sklearn