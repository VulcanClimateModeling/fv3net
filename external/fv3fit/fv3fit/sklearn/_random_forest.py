from typing import Mapping
import logging
import io
import numpy as np
from copy import copy
import xarray as xr
import pandas as pd
import fsspec
import joblib
from .._shared import (
    pack,
    unpack,
    Predictor,
    get_scaler,
    register_training_function,
)
from .._shared.config import RandomForestHyperparameters
from .. import _shared
from .._shared import (
    scaler,
    StackedBatches,
    stack_non_vertical,
    match_prediction_to_input_coords,
    SAMPLE_DIM_NAME,
)
import sklearn.base
import sklearn.ensemble

from typing import Optional, Iterable, Sequence
import yaml
from vcm import safe


def _multiindex_to_tuple(index: pd.MultiIndex) -> tuple:
    return list(index.names), list(index.to_list())


def _tuple_to_multiindex(d: tuple) -> pd.MultiIndex:
    names, list_ = d
    return pd.MultiIndex.from_tuples(list_, names=names)


@register_training_function("sklearn_random_forest", RandomForestHyperparameters)
def train_random_forest(
    hyperparameters: RandomForestHyperparameters,
    train_batches: Sequence[xr.Dataset],
    validation_batches: Sequence[xr.Dataset],
):
    model = RandomForest(
        "sample",
        hyperparameters.input_variables,
        hyperparameters.output_variables,
        hyperparameters,
    )
    # TODO: make use of validation_batches to report validation loss
    model.fit(train_batches)
    return model


@_shared.io.register("sklearn")
class RandomForest(Predictor):
    def __init__(
        self,
        sample_dim_name: str,
        input_variables: Iterable[str],
        output_variables: Iterable[str],
        hyperparameters: RandomForestHyperparameters,
    ):
        super().__init__(sample_dim_name, input_variables, output_variables)
        batch_regressor = _RegressorEnsemble(
            sklearn.ensemble.RandomForestRegressor(
                # n_jobs != 1 is non-reproducible,
                # None uses joblib which is reproducible
                n_jobs=None,
                random_state=hyperparameters.random_state,
                n_estimators=hyperparameters.n_estimators,
                max_depth=hyperparameters.max_depth,
                min_samples_split=hyperparameters.min_samples_split,
                min_samples_leaf=hyperparameters.min_samples_leaf,
                max_features=hyperparameters.max_features,
            ),
            # pass n_jobs along here to use joblib
            n_jobs=hyperparameters.n_jobs,
        )
        self._model_wrapper = SklearnWrapper(
            sample_dim_name,
            input_variables,
            output_variables,
            model=batch_regressor,
            scaler_type=hyperparameters.scaler_type,
            scaler_kwargs=hyperparameters.scaler_kwargs,
        )
        self.sample_dim_name = sample_dim_name
        self.input_variables = self._model_wrapper.input_variables
        self.output_variables = self._model_wrapper.output_variables

    def fit(self, batches: Sequence[xr.Dataset]):
        return self._model_wrapper.fit(batches)

    def predict(self, features):
        return self._model_wrapper.predict(features)

    def dump(self, path: str) -> None:
        """Dump data to a directory

        Args:
            path: a URL pointing to a directory
        """
        self._model_wrapper.dump(path)

    @classmethod
    def load(cls, path: str) -> "SklearnWrapper":
        return SklearnWrapper.load(path)


class _RegressorEnsemble:
    """
    Ensemble of RandomForestRegressor objects that are each trained on a separate
    batch of data.
    """

    def __init__(
        self,
        base_regressor,
        n_jobs,
        regressors: Sequence[sklearn.base.BaseEstimator] = None,
    ) -> None:
        self.base_regressor = base_regressor
        self.regressors = regressors or []
        self.n_jobs = n_jobs

    @property
    def n_estimators(self):
        return len(self.regressors)

    def fit(self, features, outputs):
        """ Adds a base regressor fit on features to the ensemble

        Args:
            features: numpy array of features
            outputs: numpy array of targets

        Returns:

        """
        new_regressor = copy(self.base_regressor)
        # each regressor needs different randomness
        if hasattr(new_regressor, "random_state"):
            new_regressor.random_state += len(self.regressors)
        # loky is the process-based backend
        with joblib.parallel_backend("loky", n_jobs=self.n_jobs):
            new_regressor.fit(features, outputs)
        self.regressors.append(new_regressor)

    def predict(self, features):
        """
        Args:
            features: 2D numpy array of features to predict on

        Returns:
            2D numpy array of predictions with N rows corresponding to N input samples.
            Each row is the average ensemble prediction for that sample.
        """
        predictions = np.array(
            [regressor.predict(features) for regressor in self.regressors]
        )
        return np.mean(predictions, axis=0)

    def dumps(self) -> bytes:
        batch_regressor_components = {
            "regressors": self.regressors,
            "base_regressor": self.base_regressor,
            "n_jobs": self.n_jobs,
        }
        f = io.BytesIO()
        joblib.dump(batch_regressor_components, f)
        return f.getvalue()

    @classmethod
    def loads(cls, b: bytes) -> "_RegressorEnsemble":
        f = io.BytesIO(b)
        batch_regressor_components = joblib.load(f)
        regressors: Sequence[sklearn.base.BaseEstimator] = batch_regressor_components[
            "regressors"
        ]
        base_regressor = batch_regressor_components["base_regressor"]
        obj = cls(
            base_regressor=base_regressor,
            regressors=regressors,
            n_jobs=batch_regressor_components.get("n_jobs", 1),
        )
        return obj


class SklearnWrapper(Predictor):
    """Wrap a SkLearn model for use with xarray

    """

    _PICKLE_NAME = "sklearn.pkl"
    _SCALER_NAME = "scaler.bin"
    _METADATA_NAME = "metadata.bin"

    def __init__(
        self,
        sample_dim_name: str,
        input_variables: Iterable[str],
        output_variables: Iterable[str],
        model: _RegressorEnsemble,
        scaler_type: str = "standard",
        scaler_kwargs: Optional[Mapping] = None,
    ) -> None:
        """
        Initialize the wrapper

        Args:
            sample_dim_name: dimension over which samples are taken
            input_variables: list of input variables
            output_variables: list of output variables
            model: a scikit learn regression model
        """
        self.model = model
        self.scaler_type = scaler_type
        self.scaler_kwargs = scaler_kwargs or {}
        self.target_scaler: Optional[scaler.NormalizeTransform] = None
        # TODO: remove internal sample dim name once sample dim is hardcoded everywhere
        self._input_variables = input_variables
        self._output_variables = output_variables

    def __repr__(self):
        return "SklearnWrapper(\n%s)" % repr(self.model)

    def _fit_batch(self, data: xr.Dataset):
        # TODO the sample_dim can change so best to use feature dim to flatten
        x, _ = pack(data[self.input_variables], SAMPLE_DIM_NAME)
        y, self.output_features_ = pack(data[self.output_variables], SAMPLE_DIM_NAME)

        if self.target_scaler is None:
            self.target_scaler = self._init_target_scaler(data)

        y = self.target_scaler.normalize(y)
        self.model.fit(x, y)

    def _init_target_scaler(self, batch):
        return get_scaler(
            self.scaler_type,
            self.scaler_kwargs,
            batch,
            self._output_variables,
            SAMPLE_DIM_NAME,
        )

    def fit(self, batches: Sequence[xr.Dataset]):
        logger = logging.getLogger("SklearnWrapper")
        random_state = np.random.RandomState(np.random.get_state()[1][0])
        stacked_batches = StackedBatches(batches, random_state)
        for i, batch in enumerate(stacked_batches):
            logger.info(f"Fitting batch {i+1}/{len(batches)}")
            self._fit_batch(batch)
            logger.info(f"Batch {i+1} done fitting.")

    def _predict_on_stacked_data(self, stacked_data):
        X, _ = pack(stacked_data[self.input_variables], SAMPLE_DIM_NAME)
        y = self.model.predict(X)
        if self.target_scaler is not None:
            y = self.target_scaler.denormalize(y)
        else:
            raise ValueError("Target scaler not present.")
        return unpack(y, SAMPLE_DIM_NAME, self.output_features_)

    def predict(self, data):
        stacked_data = stack_non_vertical(
            safe.get_variables(data, self.input_variables)
        )

        stacked_output = self._predict_on_stacked_data(stacked_data)
        unstacked_output = stacked_output.assign_coords(
            {SAMPLE_DIM_NAME: stacked_data[SAMPLE_DIM_NAME]}
        ).unstack(SAMPLE_DIM_NAME)

        return match_prediction_to_input_coords(data, unstacked_output)

    def dump(self, path: str) -> None:
        """Dump data to a directory

        Args:
            path: a URL pointing to a directory
        """

        fs: fsspec.AbstractFileSystem = fsspec.get_fs_token_paths(path)[0]

        fs.makedirs(path, exist_ok=True)

        mapper = fs.get_mapper(path)
        mapper[self._PICKLE_NAME] = self.model.dumps()
        if self.target_scaler is not None:
            mapper[self._SCALER_NAME] = scaler.dumps(self.target_scaler).encode("UTF-8")

        metadata = [
            self.sample_dim_name,
            self.input_variables,
            self.output_variables,
            _multiindex_to_tuple(self.output_features_),
        ]

        mapper[self._METADATA_NAME] = yaml.safe_dump(metadata).encode("UTF-8")

    @classmethod
    def load(cls, path: str) -> "SklearnWrapper":
        """Load a model from a remote path"""
        mapper = fsspec.get_mapper(path)
        model = _RegressorEnsemble.loads(mapper[cls._PICKLE_NAME])

        scaler_str = mapper.get(cls._SCALER_NAME, b"")
        scaler_obj: Optional[scaler.NormalizeTransform]
        if scaler_str:
            scaler_obj = scaler.loads(scaler_str)
        else:
            scaler_obj = None
        (
            sample_dim_name,
            input_variables,
            output_variables,
            output_features_dict_,
        ) = yaml.safe_load(mapper[cls._METADATA_NAME])

        output_features_ = _tuple_to_multiindex(output_features_dict_)

        obj = cls(sample_dim_name, input_variables, output_variables, model)
        obj.target_scaler = scaler_obj
        obj.output_features_ = output_features_

        return obj

    # these are here for backward compatibility with pre-unified API attribute names
    @property
    def input_variables(self):
        if hasattr(self, "_input_variables"):
            return self._input_variables
        elif hasattr(self, "input_vars_"):
            return self.input_vars_
        else:
            raise ValueError("Wrapped model version without input variables attribute.")

    @property
    def output_variables(self):
        if hasattr(self, "_input_variables"):
            return self._output_variables
        elif hasattr(self, "input_vars_"):
            return self.output_vars_
        else:
            raise ValueError(
                "Wrapped model version without output variables attribute."
            )

    @property
    def sample_dim_name(self):
        return getattr(self, "_sample_dim_name", "sample")
