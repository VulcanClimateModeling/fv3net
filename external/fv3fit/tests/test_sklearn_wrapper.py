from external.fv3fit.fv3fit._shared import stack_non_vertical
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.dummy import DummyRegressor
import unittest.mock
import pytest
import xarray as xr
import joblib

from fv3fit.sklearn._random_forest import _RegressorEnsemble, pack, SklearnWrapper
from fv3fit._shared.scaler import ManualScaler


def test_flatten():
    x = np.ones((3, 4, 5))
    shape = (3, 4, 5)
    dims = "x y z".split()
    sample_dim = "z"

    nx, ny, nz = shape

    a = xr.DataArray(x, dims=dims)
    ds = xr.Dataset({"a": a, "b": a})

    ans = pack(ds, sample_dim)[0]
    assert ans.shape == (nz, 2 * nx * ny)


def test_flatten_1d_input():
    x = np.ones((3, 4, 5))
    shape = (3, 4, 5)
    dims = "x y z".split()
    sample_dim = "z"

    nx, ny, nz = shape

    a = xr.DataArray(x, dims=dims)
    ds = xr.Dataset({"a": a, "b": a.isel(x=0, y=0)})

    ans = pack(ds, sample_dim)[0]
    assert ans.shape == (nz, nx * ny + 1)


def test_flatten_same_order():
    nx, ny = 10, 4
    x = xr.DataArray(np.arange(nx * ny).reshape((nx, ny)), dims=["sample", "feature"])

    ds = xr.Dataset({"a": x, "b": x.T})
    sample_dim = "sample"
    a = pack(ds[["a"]], sample_dim)[0]
    b = pack(ds[["b"]], sample_dim)[0]

    np.testing.assert_allclose(a, b)


@pytest.fixture
def test_regressor_ensemble():
    base_regressor = LinearRegression()
    ensemble_regressor = _RegressorEnsemble(base_regressor, n_jobs=1)
    num_batches = 3
    X = np.array([[1, 1], [1, 2], [2, 2], [2, 3]])
    y = np.dot(X, np.array([1, 2])) + 3
    for i in range(num_batches):
        ensemble_regressor.fit(X, y)
    return ensemble_regressor


def test_ensemble_fit(test_regressor_ensemble):
    regressor_ensemble = test_regressor_ensemble
    assert regressor_ensemble.n_estimators == 3
    X = np.array([[1, 1], [1, 2], [2, 2], [2, 3]])
    y = np.dot(X, np.array([1, 2])) + 3
    regressor_ensemble.fit(X, y)
    # test that .fit appends a new regressor
    assert regressor_ensemble.n_estimators == 4
    # test that new regressors are actually fit and not empty base regressor
    assert len(regressor_ensemble.regressors[-1].coef_) > 0


def _get_sklearn_wrapper(scale_factor=None, dumps_returns: bytes = b"HEY!"):
    model = unittest.mock.Mock()
    model.regressors = []
    model.base_regressor = unittest.mock.Mock()
    model.predict.return_value = np.array([[1.0]])
    model.dumps.return_value = dumps_returns

    if scale_factor:
        scaler = ManualScaler(np.array([scale_factor]))
    else:
        scaler = None

    wrapper = SklearnWrapper(
        sample_dim_name="sample",
        input_variables=["x"],
        output_variables=["y"],
        model=model,
    )
    wrapper.target_scaler = scaler
    return wrapper


def test_SklearnWrapper_fit_predict_scaler(scale=2.0):
    wrapper = _get_sklearn_wrapper(scale)
    dims = ["unstacked_dim", "z"]
    data = xr.Dataset({"x": (dims, np.ones((1, 1))), "y": (dims, np.ones((1, 1)))})
    wrapper.fit([data])
    stacked_data = stack_non_vertical(data)
    output = wrapper.predict(stacked_data)
    assert pytest.approx(1 / scale) == output["y"].item()


def test_fitting_SklearnWrapper_does_not_fit_scaler():
    """SklearnWrapper should use pre-computed scaling factors when fitting data
    
    In other words, calling the .fit method of wrapper should not call the
    .fit its scaler attribute.
    """

    model = unittest.mock.Mock()
    scaler = unittest.mock.Mock()

    wrapper = SklearnWrapper(
        sample_dim_name="sample",
        input_variables=["x"],
        output_variables=["y"],
        model=model,
    )
    wrapper.target_scaler = scaler

    dims = ["sample_", "z"]
    data = xr.Dataset({"x": (dims, np.ones((1, 1))), "y": (dims, np.ones((1, 1)))})
    wrapper.fit([data])
    scaler.fit.assert_not_called()


@pytest.mark.parametrize(
    "scale_factor", [2.0, None],
)
def test_SklearnWrapper_serialize_predicts_the_same(tmpdir, scale_factor):

    # Setup wrapper
    if scale_factor:
        scaler = ManualScaler(np.array([scale_factor]))
    else:
        scaler = None
    model = _RegressorEnsemble(base_regressor=LinearRegression(), n_jobs=1)
    wrapper = SklearnWrapper(
        sample_dim_name="sample",
        input_variables=["x"],
        output_variables=["y"],
        model=model,
    )
    wrapper.target_scaler = scaler

    # setup input data
    dims = ["unstacked_dim", "z"]
    data = xr.Dataset({"x": (dims, np.ones((1, 1))), "y": (dims, np.ones((1, 1)))})
    wrapper.fit([data])

    # serialize/deserialize
    path = str(tmpdir)
    wrapper.dump(path)

    loaded = wrapper.load(path)
    stacked_data = stack_non_vertical(data)
    xr.testing.assert_equal(loaded.predict(stacked_data), wrapper.predict(stacked_data))


def test_SklearnWrapper_serialize_fit_after_load(tmpdir):
    model = _RegressorEnsemble(base_regressor=LinearRegression(), n_jobs=1)
    wrapper = SklearnWrapper(
        sample_dim_name="sample",
        input_variables=["x"],
        output_variables=["y"],
        model=model,
    )

    # setup input data
    dims = ["unstacked_dim", "z"]
    data = xr.Dataset({"x": (dims, np.ones((1, 1))), "y": (dims, np.ones((1, 1)))})
    wrapper.fit([data])

    # serialize/deserialize
    path = str(tmpdir)
    wrapper.dump(path)

    # fit loaded model
    loaded = wrapper.load(path)
    loaded.fit([data])

    assert len(loaded.model.regressors) == 2


def test_predict_columnwise_is_deterministic(regtest):
    """Tests that fitting/predicting with a model is deterministic

    If this fails, look for non-deterministic logic (e.g. converting sets to lists)
    """
    nz = 2
    model = _RegressorEnsemble(
        base_regressor=DummyRegressor(strategy="constant", constant=np.arange(nz)),
        n_jobs=1,
    )
    wrapper = SklearnWrapper(
        sample_dim_name="sample",
        input_variables=["a"],
        output_variables=["b"],
        model=model,
    )

    dims = ["x", "y", "z"]
    shape = (2, 2, nz)
    arr = np.arange(np.prod(shape)).reshape(shape)
    data = xr.Dataset({"a": (dims, arr), "b": (dims, arr + 1)})
    wrapper.fit([data])

    output = wrapper.predict_columnwise(data, feature_dim="z")
    print(joblib.hash(np.asarray(output["b"])), file=regtest)
