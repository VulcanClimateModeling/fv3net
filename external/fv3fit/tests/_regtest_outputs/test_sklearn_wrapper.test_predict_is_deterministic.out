96637febbfefdf9f34989e960043c6d8
